### PR TITLE
feat(grid-esf): make excel style filtering work like in excel

### LINKED ISSUE #5812: IgxGrid: Excel Style Filter "Apply" on "Enter"
## Is your feature request related to a problem? Please describe.
When I open the Excel Style Filter on a column I would like to just search and hit the "Enter"-key to apply the changes.

## Describe the solution you'd like
When Excel Style Filter is open and the user hits "Enter" the selection should be applied

### LINKED ISSUE #5813: IgxGrid: Excel Style Filter "Search" as in Excel
## Is your feature request related to a problem? Please describe.
In a Excel Style Filter I can "Search" as in Excel. However the behavior is different. I first need to uncheck "Select All" and then to recheck the items I want to see in the grid.
It would be great to create a shortcut for that.

## Describe the solution you'd like
It would be nice to have it implemented as in Excel.
1. Type in your Search-Text
2. Hit "Enter" or click "Apply"
3. Only the rows that match your Search-Text are displayed

### LINKED ISSUE #6617: Incorrect behavior of the Select All checkbox in Excel Style Filtering
## Description  
When a user types a value in the search input of the Excel Style Filtering, items get filtered. Then, if a user unchecks all filtered items (except for Select All checkbox), the Select All checkbox goes indetermined, instead of unchecked. Select All affect all items, even if they are filtered by the value in the search input. There is no way to act on the filtered items all at once.

## Steps to reproduce  

1.  The goal is to filter all users by the name containing "ste"
2. Trigger Excel Style Filtering
3. Deselect all items by unchecking the "Select All" item in the list of unique values
2. Type "ste" in the search input
3. It is not possible to select just the filtered items as "Select All" would check all items in the grid, independent from the search term typed in

## Result  
Finally, Select All goes indetermined instead of unchecked. 

Screen Shot 2020-02-05 at 3 26 51 PM

## Expected result  
I expect to be able to select/deselect all filtered items at once when a search criterion is specified. 

## Solution
- When having a value in the search input, we will add not one but two items besides the unique ones
- First will appear: "Select All items" - affecting all list items
- Next will come: "Select Filtered Items" - affecting only the filtered list items

### PR DESCRIPTION
**Pull request by onlyexeption**
Closes #5813 
Closes #6617 
Closes #5812

### Additional information (check all that apply):
 - [ ] Bug fix
 - [x] New functionality
 - [ ] Documentation
 - [ ] Demos
 - [ ] CI/CD

### Checklist:
 - [x] All relevant tags have been applied to this PR
 - [x] This PR includes unit tests covering all the new code ([test guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Test-implementation-guidelines-for-Ignite-UI-for-Angular))
 - [ ] This PR includes API docs for newly added methods/properties ([api docs guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Documentation-Guidelines))
 - [ ] This PR includes `feature/README.MD` updates for the feature docs
 - [ ] This PR includes general feature table updates in the root `README.MD`
 - [x] This PR includes `CHANGELOG.MD` updates for newly added functionality
 - [ ] This PR contains breaking changes
 - [ ] This PR includes `ng update` migrations for the breaking changes ([migrations guidelines](https://github.com/IgniteUI/igniteui-angular/wiki/Update-Migrations))
 - [ ] This PR includes behavioral changes and the feature specification has been updated with them

### CODE REVIEW

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts [hunk 1]
```diff
@@ -203,12 +223,77 @@ export class IgxExcelStyleSearchComponent implements AfterViewInit, OnDestroy {
...
+        }
+
+        const searchVal = this.searchValue.toLowerCase();
+
+        this.displayedListData = this.esf.listData.filter((it, i) => (i === 0 && it.isSpecial) ||
+            (it.label !== null && it.label !== undefined) &&
+            !it.isBlanks &&
+            it.label.toString().toLowerCase().indexOf(searchVal) > -1);
+
+        this.esf.listData.forEach(i => this.displayedListData.indexOf(i) === -1 ? i.isSelected = false : i.isSelected = true);
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts:273]: We should optimize this.

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts [hunk 2]
```diff
@@ -262,6 +347,8 @@ export class IgxExcelStyleSearchComponent implements AfterViewInit, OnDestroy {
             this.esf.expressionsList = new Array<ExpressionUI>();
             this.esf.grid.filteringService.generateExpressionsList(this.esf.column.filteringExpressionsTree,
                 this.esf.grid.filteringLogic, this.esf.expressionsList);
+
+            this.esf.listData.forEach(i => i.isSelected ? i.isFiltered = true : i.isFiltered = false);
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts:351]:
```suggestion
            this.esf.listData.forEach(i => i.isFiltered = i.isSelected);
```

**Review by DiyanDimitrov — CHANGES_REQUESTED** [review]
Add the feature to the changelog.

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts [hunk 3]
```diff
@@ -154,28 +174,40 @@ export class IgxExcelStyleSearchComponent implements AfterViewInit, OnDestroy {
...
      */
     public onCheckboxChange(eventArgs: IChangeCheckboxEventArgs) {
-        const selectedIndex = this.esf.listData.indexOf(eventArgs.checkbox.value);
+        const selectedIndex = this.displayedListData.indexOf(eventArgs.checkbox.value);
+        const selectAllBtn = this.displayedListData[0];
+
         if (selectedIndex === 0) {
-            this.esf.listData.forEach(element => {
+            this.displayedListData.forEach(element => {
+                if (element.value === this.addToCurrentFilter.value) { return; }
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts:189]: This should work with comparing the objects themselves.
```suggestion
                if (element === this.addToCurrentFilter) { return; }
```

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts [hunk 4]
```diff
@@ -154,28 +167,35 @@ export class IgxExcelStyleSearchComponent implements AfterViewInit, OnDestroy {
...
             eventArgs.checkbox.value.isSelected = eventArgs.checked;
-            if (!this.esf.listData.slice(1, this.esf.listData.length).find(el => el.isSelected === false)) {
-                this.esf.listData[0].indeterminate = false;
-                this.esf.listData[0].isSelected = true;
-            } else if (!this.esf.listData.slice(1, this.esf.listData.length).find(el => el.isSelected === true)) {
-                this.esf.listData[0].indeterminate = false;
-                this.esf.listData[0].isSelected = false;
+            const indexToStartSlicing = this.displayedListData.indexOf(this.addToCurrentFilter) > -1 ? 2 : 1;
+
+            if (!this.displayedListData.slice(indexToStartSlicing, this.displayedListData.length).find(el => el.isSelected === false)) {
```
[the remark this answers is not in the document]
Comment (DiyanDimitrov) [reply projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts]: The array is still sliced two times. I have meant to have:
```ts
const slicedArray = this.displayedListData.slice(indexToStartSlicing, this.displayedListData.length);
```

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts [hunk 5]
```diff
@@ -203,12 +223,77 @@ export class IgxExcelStyleSearchComponent implements AfterViewInit, OnDestroy {
...
+
+            return;
+        }
+
+        if (!this.searchValue) {
+            const anyFiltered = this.esf.listData.some(i => i.isFiltered);
+
+            if (anyFiltered) {
+                this.esf.listData.forEach(i => i.isSelected = i.isFiltered);
+                this.displayedListData[0].indeterminate = true;
```
[the remark this answers is not in the document]
Comment (DiyanDimitrov) [reply projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts]: This issue still persists. You could see a console error when typing a search value that doesn't match any items and then delete it.

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts (hunk 2 above)
Comment (DiyanDimitrov) [reply projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts]: The code now is `i.isSelected = i.isFiltered`, shouldn't it be the opposite?

File: CHANGELOG.md [hunk 6]
```diff
@@ -4,6 +4,12 @@ All notable changes for each version of this project will be documented in this
 ## 10.2.0
 
 ### General
+- `IgxGrid`, `IgxTreeGrid`, `IgxHierarchicalGrid`
+    - **Behavioral Change** - The Excel Style Filtering has been reworked to provide filtering experience such as in Excel. This includes the following changes.
+        - You can close the Excel Style Filtering menu by pressing `Ctrl + Shift + L`.
+        - You can apply the filter by pressing `Enter`.
+        - When filtering, only the rows that match your search term will be visible.
```
Comment (DiyanDimitrov) [inline CHANGELOG.md:11]:
```suggestion
        - When searching items in the Excel Style Filtering menu, only the rows that match your search term will be filtered in.
```

File: CHANGELOG.md [hunk 7]
```diff
@@ -4,6 +4,12 @@ All notable changes for each version of this project will be documented in this
 ## 10.2.0
 
 ### General
+- `IgxGrid`, `IgxTreeGrid`, `IgxHierarchicalGrid`
+    - **Behavioral Change** - The Excel Style Filtering has been reworked to provide filtering experience such as in Excel. This includes the following changes.
```
Comment (DiyanDimitrov) [inline CHANGELOG.md:8]:
```suggestion
    - **Behavioral Change** - The Excel Style Filtering has been reworked to provide filtering experience such as in Excel. This includes the following changes:
```

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts [hunk 8]
```diff
@@ -124,11 +139,16 @@ export class IgxExcelStyleSearchComponent implements AfterViewInit, OnDestroy {
         esf.initialized.pipe(takeUntil(this.destroy$)).subscribe(() => {
             requestAnimationFrame(() => {
                 this.searchInput.nativeElement.focus();
+
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts:142]: Remove empty line.

**Review by DiyanDimitrov — CHANGES_REQUESTED** [review]

File: projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts [hunk 9]
```diff
@@ -203,12 +232,80 @@ export class IgxExcelStyleSearchComponent implements AfterViewInit, OnDestroy {
...
+    /**
+     * @hidden @internal
+     */
+    get applyButtonDisabled(): boolean {
+        return this.esf.listData[0] && !this.esf.listData[0].isSelected && !this.esf.listData[0].indeterminate ||
+            this.displayedListData && this.displayedListData.length === 0;
+    }
+
+
+    public onInputKeyDown(event): void {
```
Comment (DiyanDimitrov) [inline projects/igniteui-angular/src/lib/grids/filtering/excel-style/excel-style-search.component.ts:245]: Add `@hidden` and `@internal` to all public members.

**Review by DiyanDimitrov — CHANGES_REQUESTED** [review]

**Review by DiyanDimitrov — APPROVED** [review]

File: projects/igniteui-angular/src/lib/grids/grid/grid-filtering-ui.spec.ts [hunk 10]
```diff
@@ -1641,25 +1641,32 @@ describe('IgxGrid - Filtering Row UI actions #grid', () => {
...
+                const headerGroupsFiltering = headerGroups.filter(
+                    (hg) => hg.nativeElement.classList.contains('igx-grid__th--filtering'));
+                expect(headerGroupsFiltering.length).toBe(1);
+                expect(headerGroupsFiltering[0].componentInstance.column.field).toBe('ReleaseDate');
+
+                UIInteractions.triggerKeyDownEvtUponElem('l', releaseDateColumn.nativeElement, true, false, true, true);
+                tick(200);
+                fix.detectChanges();
+
+                expect(fix.debugElement.query(By.css(FILTER_UI_ROW))).not.toBeNull();
```
Comment (igdmdimitrov) [inline projects/igniteui-angular/src/lib/grids/grid/grid-filtering-ui.spec.ts:1669]: Ctrl+Shift+L should close the Filter Row and the last statement will change

File: projects/igniteui-angular/src/lib/grids/grid/grid-filtering-ui.spec.ts [hunk 11]
```diff
@@ -4344,7 +4353,195 @@ describe('IgxGrid - Filtering actions - Excel style filtering #grid', () => {
...
+        tick(100);
+        fix.detectChanges();
+
+        // Get the results and verify that they match the list items.
+        gridCellValues = GridFunctions.getColumnCells(fix, 'Downloads')
+            .map(c => c.nativeElement.innerText)
+            .sort();
+
+        expect(gridCellValues.length).toEqual(3);
+        expect(gridCellValues).toEqual(totalListItems);
```
Comment (igdmdimitrov) [inline projects/igniteui-angular/src/lib/grids/grid/grid-filtering-ui.spec.ts:4511]: If possible - at the end of the test open the filter dialog again and verify the selection state of the 'Select All' button.

**Review by DiyanDimitrov — APPROVED** [review]

**Review by DiyanDimitrov — APPROVED** [review]

**Review by DiyanDimitrov — APPROVED** [review]

**Pull request merged**